### ISSUE TITLE
What do you think about extending Multer with a few new options?

### ISSUE THREAD

**Issue by turakvlad** [expressjs/multer#646]
What do you think about extending Multer with a few new options such as `minCount` and `count` as an addition to existing `maxCount`? 

Honestly speaking, I need these options. For now, I can check only `maxCount`. If my application needs to get at least three pictures from a client, I cannot check this with Multer. For now, Joi handles it. I don't like to separate these validations mechanism. Validating at least the basic things with Multer would be great.

Also, we can add `fileType` option. We can use [this](https://www.npmjs.com/package/file-type) library to check the files' types. I understand that this is possible to do with the Multer's `fileFilter` function. But I think it would be great to have this filter. It will help the new users to enter into Multer faster. They will have a few useful options out of the box.

@LinusU, I will be happy to hear your thoughts.

**Comment by Matthew-Herman** [comment]
A fileType option would benice for convenience though it's fairly trivial to implement by checking mimetypes in filefilter.

I think a minCount option would be nice, though it seems to be related to a feature that is really lacking from multer.

The feature that is lacking is to add middleware that can process the input of all other fields besides files, before filefilter runs and before the files are uploaded.

This would require delaying the file uploads until all the fields in the form have been processed, and only after that upload the files. 

This would be really useful because it seems to be a really common case that if you have a form with multiple files and text inputs, all of which need to be validated, you wouldn't want to upload the files before you've checked the non-file inputs, because this results in wastefully having to remove the uploaded files if any of the non-file inputs were invalid.

It's really unfortunate that with multer alone, you cannot access the non-file data of multipart forms before running the middleware returned by calling multer.fields.

**Comment by LinusU** [comment]
@turakvlad first of all, sorry that I never got back to you on this, I've must have missed the notification.

The reason that we only provide `maxCount` is since the use case here is only to be able to terminate early as to prevent DOS attacks, it's not intended as a form validation library.

@Matthew-Herman have you looked at the alpha for 2.x (see #399), I think that that might be in line with what you are trying to do 👍

**Comment by turakvlad** [comment]
@LinusU, hey! No worries at all! :)

So, if you think that we don't need these options at all, we can close this issue.

**Comment by shierro** [comment]
Hi, i also think that the `minCount` option would be very useful. i have a case, where I upload multiple files via `.fields([...array])` 

I wanted to check if those files exist, before sending it to s3 via `multer-s3` 

Was looking for something like `{ name: 'avatar', maxCount: 1, minCount: 1 }` or `{ name: 'avatar', maxCount: 1, required: true }`

and landed on this issue :D

**Comment by ryhinchey** [comment]
> The reason that we only provide `maxCount` is since the use case here is only to be able to terminate early as to prevent DOS attacks, it's not intended as a form validation library.

I'm going to go ahead and close this issue per @LinusU comment.  Implementing further checks around the number of files being uploaded is currently out of scope for multer.
